akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -354,16 +354,75 @@ final class CoordinatedShutdown private[akka] (
     system: ExtendedActorSystem,
     phases: Map[String, CoordinatedShutdown.Phase])
     extends Extension {
-  import CoordinatedShutdown.Reason
-  import CoordinatedShutdown.UnknownReason
+  import CoordinatedShutdown.{ Reason, UnknownReason }
 
   /** INTERNAL API */
   private[akka] val log = Logging(system, getClass)
   private val knownPhases = phases.keySet ++ phases.values.flatMap(_.dependsOn)
 
   /** INTERNAL API */
   private[akka] val orderedPhases = CoordinatedShutdown.topologicalSort(phases)
-  private val tasks = new ConcurrentHashMap[String, Vector[(String, () => Future[Done])]]
+
+  private object tasks {
+    private val registeredTasks = new ConcurrentHashMap[String, Map[() => Future[Done], Vector[String]]]()
+
+    def get(phase: String): Option[Map[() => Future[Done], Vector[String]]] = Option(registeredTasks.get(phase))
+
+    def totalDuration(): FiniteDuration = {
+      import akka.util.ccompat.JavaConverters._
+      registeredTasks.keySet.asScala.foldLeft(Duration.Zero) {
+        case (acc, phase) =>
+          acc + timeout(phase)
+      }
+    }
+
+    def register(phase: String)(task: () => Future[Done], name: String): Cancellable = {
+      registeredTasks.merge(
+        phase,
+        Map(task -> Vector(name)),
+        (t, u) =>
+          t ++ u.map {
+            case (k, names) =>
+              k -> (t.get(k).toVector.flatten ++ names)
+          })
+      new Cancellable {
+        private val p = Promise[Unit]()
+
+        override def cancel(): Boolean = {
+          if (p.isCompleted) { // This ensures that multiple calls to `cancel` don't unregister another task with the same name
+            false
+          } else {
+            unregister(phase)(task, name) && p.trySuccess {
+              log.info("Successfully unregistered task [{}] from phase {}.", name, phase)
+            }
+          }
+        }
+
+        override def isCancelled: Boolean = p.isCompleted
+      }
+    }
+
+    private def unregister(phase: String)(task: () => Future[Done], name: String): Boolean = {
+      Option(registeredTasks.get(phase)).fold(false) { phaseDef =>
+        phaseDef.get(task).fold(false) {
+          case names if names.contains(name) =>
+            registeredTasks.merge(
+              phase,
+              Map(task -> Vector(name)),
+              (prev, subt) =>
+                (prev ++ subt.flatMap {
+                  case (t, ns) =>
+                    prev.get(t).map { prevNames =>
+                      t -> prevNames.diff(ns) // replaces the previous job names at t with a sequence with one fewer occurrence of each name in `ns`
+                    }
+                }).filter(_._2.nonEmpty))
+            true
+          case _ =>
+            false
+        }
+      }
+    }
+  }
```

Comment:
I'm not fully convinced that it's good to add this complexity for this feature. It was not the original purpose of CoordinatedShutdown to use it as a dynamic registry of many tasks. Wonder if it can be simplified by providing something like a "composite task" that is registered once (with existing API) to CoordinatedShutdown and then the dynamic sub-tasks can be added/removed to that?

Reply:
I can see two things you might be referring to with "complexity" above.

1. The change to the public API. I don't think this API change is terribly complex, but I'm open to being persuaded otherwise.

2. The internal plumbing to get cancellation to work. Here I feel pretty strongly that "adding complexity" is not a fair description. As it stands in master, the code that adds a task (which I think you wrote if I'm reading the `git blame` correctly) contains all this complexity in a terser form by bypassing the higher-level methods of `ConcurrentHashMap` in favor of rolling our own atomic update with explicit tail recursion, requiring handling two possible `null` return values and interpreting a boolean return value in the second branch.  (If you think the logic could be made easier to read by, e.g., better variable naming, I'm happy to oblige there.)

Reply:
ok, we can work on making the internals easier to follow with better naming, but let's first define the desired semantics of the API.

Reply:
Agreed